output/cloud/expv2/output.go

```diff
@@ -0,0 +1,265 @@
+// binary format for encoding payloads.
+package expv2
+
+import (
+	"context"
+	"time"
+
+	"github.com/mstoykov/atlas"
+	"github.com/sirupsen/logrus"
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/metrics"
+	"go.k6.io/k6/output"
+	"go.k6.io/k6/output/cloud/expv2/pbcloud"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// TestName is the default Cloud test name
+const (
+	TestName = "k6 test"
+)
+
+// Output sends result data to the k6 Cloud service.
+type Output struct {
+	output.SampleBuffer
+
+	config      cloudapi.Config
+	referenceID string
+
+	logger          logrus.FieldLogger
+	client          *MetricsClient
+	periodicFlusher *output.PeriodicFlusher
+
+	// TODO: if the metric refactor (#2905) will introduce
+	// a sequential ID for metrics
+	// then we could reuse the same strategy here
+	activeSeries map[*metrics.Metric]aggregatedSamples
+}
+
+// New creates a new cloud output.
+func New(logger logrus.FieldLogger, conf cloudapi.Config) (*Output, error) {
+	mc, err := NewMetricsClient(logger, conf.Host.String, conf.Token.String)
+	if err != nil {
+		return nil, err
+	}
+	return &Output{
+		config:       conf,
+		client:       mc,
+		logger:       logger.WithFields(logrus.Fields{"output": "cloudv2"}),
+		activeSeries: make(map[*metrics.Metric]aggregatedSamples),
+	}, nil
+}
+
+// Start starts the output.
+func (o *Output) Start() error {
+	o.logger.Debug("Starting...")
+
+	// TODO: merge here the part executed by v1 when we will drop it
+	pf, err := output.NewPeriodicFlusher(o.config.MetricPushInterval.TimeDuration(), o.flushMetrics)
+	if err != nil {
+		return err
+	}
+	o.logger.Debug("Started!")
+	o.periodicFlusher = pf
+	return nil
+}
+
+// StopWithTestError stops the output.
+func (o *Output) StopWithTestError(testErr error) error {
+	o.logger.Debug("Stopping...")
+	defer o.logger.Debug("Stopped!")
+	o.periodicFlusher.Stop()
+	return nil
+}
+
+// SetReferenceID sets the Cloud's test ID.
+func (o *Output) SetReferenceID(refID string) {
+	o.referenceID = refID
+}
+
+// AddMetricSamples receives a set of metric samples.
+func (o *Output) flushMetrics() {
+	if o.referenceID == "" {
+		// TODO: should it warn?
+		return
+	}
+
+	start := time.Now()
+
+	if hasOne := o.collectSamples(); !hasOne {
+		return
+	}
+
+	// TODO: in case an aggregation period will be added then
+	// it continue only if the aggregation time frame passed
+
+	metricSet := make([]*pbcloud.Metric, 0, len(o.activeSeries))
+	for m, aggr := range o.activeSeries {
+		if len(aggr.Samples) < 1 {
+			// If a bucket (a metric) has been added
+			// then the assumption is to collect at least once in a flush interval.
+			continue
+		}
+		metricSet = append(metricSet, o.mapMetricProto(m, aggr))
+		aggr.Clean()
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), o.config.MetricPushInterval.TimeDuration())
+	defer cancel()
+
+	err := o.client.Push(ctx, o.referenceID, &pbcloud.MetricSet{Metrics: metricSet})
+	if err != nil {
+		o.logger.Error(err)
```

Comment:
Let's be more precise here (instead of prefixing some of the errors inside the `Push` method with `failed to push metrics:`). 

That way we avoid any bare/unwrapped error from inside.